### PR TITLE
Add a separate Group State for critical events

### PR DESCRIPTION
**Pull request by AnastasiaKubova**
Please have a look at our [guidelines for contributions](https://github.com/Microsoft/AppCenter-SDK-Android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Add an instance of GroupState for critical events in Analytics (using addGroup), in the same way as it's done in OneCollector. Name it serviceName + "Critical" (this should be Analytics/Critical).
Implement adding events to a different group depending on the priority.

## Related PRs or issues

[AB#61308](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/61308)

### CODE REVIEW

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 1]
```diff
@@ -48,21 +50,19 @@
  */
 public class Analytics extends AbstractAppCenterService {
 
+    public static final String ANALYTICS_CRITICAL_GROUP = "group_critical";
+    /**
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]: Add empty line

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 2]
```diff
@@ -48,21 +50,19 @@
  */
 public class Analytics extends AbstractAppCenterService {
 
+    public static final String ANALYTICS_CRITICAL_GROUP = "group_critical";
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]:
```suggestion
    public static final String ANALYTICS_CRITICAL_GROUP = ANALYTICS_GROUP + "_critical";
```

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 3]
```diff
@@ -48,21 +50,19 @@
...
+    /**
+     * Constant marking event of the analytics group.
+     */
+    static final String ANALYTICS_GROUP = "group_analytics";
     /**
      * Name of the service.
      */
     private static final String SERVICE_NAME = "Analytics";
-
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]: Restore empty line

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 4]
```diff
@@ -48,21 +50,19 @@
...
     /**
      * TAG used in logging for Analytics.
      */
     public static final String LOG_TAG = AppCenterLog.LOG_TAG + SERVICE_NAME;
-
-    /**
-     * Constant marking event of the analytics group.
-     */
-    static final String ANALYTICS_GROUP = "group_analytics";
-
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]: Restore empty line

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 5]
```diff
@@ -785,11 +787,19 @@ public void run() {
...
                 /* Filter and validate flags. For now we support only persistence. */
                 int filteredFlags = Flags.getPersistenceFlag(flags, true);
-                mChannel.enqueue(eventLog, ANALYTICS_GROUP, filteredFlags);
+                mChannel.enqueue(eventLog, isCritical(filteredFlags) ? ANALYTICS_CRITICAL_GROUP : ANALYTICS_GROUP, filteredFlags);
             }
         });
     }
 
+    private boolean isCritical(int flag) {
+        return flag == Flags.PERSISTENCE_CRITICAL;
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]:
```suggestion
        return flag & Flags.PERSISTENCE_CRITICAL != 0;
```

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 6]
```diff
@@ -785,11 +787,19 @@ public void run() {
...
+    }
+
+    private String criticalAnalyticsGroupName(String groupName) {
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]: Remove this method

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 7]
```diff
@@ -656,13 +656,15 @@ protected synchronized void applyEnabledState(boolean enabled) {
...
         if (enabled) {
+            mChannel.addGroup(criticalAnalyticsGroupName("Critical"), getTriggerCount(), getTriggerInterval(), getTriggerMaxParallelRequests(), null, getChannelListener());
 
             /* Check if service started at application level and enable corresponding features. */
             startAppLevelFeatures();
         }
 
         /* On disabling service. */
         else {
+            mChannel.removeGroup(criticalAnalyticsGroupName("Critical"));
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]:
```suggestion
            mChannel.removeGroup(ANALYTICS_CRITICAL_GROUP);
```

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 8]
```diff
@@ -656,13 +656,15 @@ protected synchronized void applyEnabledState(boolean enabled) {
 
         /* If we enabled the service. */
         if (enabled) {
+            mChannel.addGroup(criticalAnalyticsGroupName("Critical"), getTriggerCount(), getTriggerInterval(), getTriggerMaxParallelRequests(), null, getChannelListener());
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]:
```suggestion
            mChannel.addGroup(ANALYTICS_CRITICAL_GROUP, getTriggerCount(), getTriggerInterval(), getTriggerMaxParallelRequests(), null, getChannelListener());
```

Replace `getTriggerInterval` to new const

File: sdk/appcenter-analytics/src/test/java/com/microsoft/appcenter/analytics/AnalyticsTest.java [hunk 9]
```diff
@@ -79,6 +80,16 @@
 
 public class AnalyticsTest extends AbstractAnalyticsTest {
 
+    /**
+     * Constant marking event of the analytics group.
+     */
+    static final String ANALYTICS_GROUP = "group_analytics";
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/test/java/com/microsoft/appcenter/analytics/AnalyticsTest.java]: Remove

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 10]
```diff
@@ -656,13 +663,15 @@ protected synchronized void applyEnabledState(boolean enabled) {
 
         /* If we enabled the service. */
         if (enabled) {
+            mChannel.addGroup(ANALYTICS_CRITICAL_GROUP, getTriggerCount(), getTriggerInterval(), getTriggerMaxParallelRequests(), null, getChannelListener());
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]: Use different interval

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java [hunk 11]
```diff
@@ -785,11 +794,15 @@ public void run() {
 
                 /* Filter and validate flags. For now we support only persistence. */
                 int filteredFlags = Flags.getPersistenceFlag(flags, true);
-                mChannel.enqueue(eventLog, ANALYTICS_GROUP, filteredFlags);
+                mChannel.enqueue(eventLog, isCritical(filteredFlags) ? ANALYTICS_CRITICAL_GROUP : ANALYTICS_GROUP, filteredFlags);
```
Comment (MatkovIvan) [inline sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]: Can we reuse `getPersistenceFlag` here?

File: sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java (hunk 10 above)
Comment (annakocheshkova) [reply sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/Analytics.java]: 3.0 for critical events

File: sdk/appcenter-analytics/src/test/java/com/microsoft/appcenter/analytics/AnalyticsTest.java [hunk 12]
```diff
@@ -490,7 +495,8 @@ public void setEnabled() throws InterruptedException {
         verify(channel).removeListener(isA(SessionTracker.class));
         verify(channel).removeListener(isA(AnalyticsValidator.class));
         verify(channel).removeListener(isA(AnalyticsTransmissionTarget.getChannelListener().getClass()));
-        verify(channel, times(2)).removeGroup(analytics.getGroupName());
+        verify(channel).removeGroup(eq(ANALYTICS_CRITICAL_GROUP));
+        verify(channel, times(2)).removeGroup(eq(ANALYTICS_GROUP));
```
Comment (guperrot) [inline sdk/appcenter-analytics/src/test/java/com/microsoft/appcenter/analytics/AnalyticsTest.java]: why is it asymmetric between the 2 groups? they should have the same life cycle, not sure to understand what is going on here

Comment (guperrot) [reply sdk/appcenter-analytics/src/test/java/com/microsoft/appcenter/analytics/AnalyticsTest.java]: Please ignore my comment I remember where this comes from, abstract code removes the group before adding it.

**Pull request merged**